feat: added ability to receive a base64 with compressed clientstate, #6918

goto40 opened the pull request:
I encountered a problem with large URLs when using the "clientstate" API with base64 strings: see #6918 

* I found that it was quite easy to include this feature, just by searching for `/clientstate/`.
* I "tested" my feature interactively with `npm run dev` and this [link](http://localhost:10240/zclientstate/eNq1VduK2zAQ/ZXBhWITJc6lVye7sBT6uIWl9CUOiyJPYm1syUjyZsOSf+/Idm5tWuhD7YCs0cyZM5doXgOL1kqtbJDA/DWQGa0jBkHB1brma6RtIHq9gERW10Y0glTFMYiqMrhCg0pgArlzlU3iGNXg/GAgdBlvYxLFQivHpUITl5mtuErVG6lEUWcIM6mtM8jL23PhAdLw7WAtXV4va4vGw6ByDfBGbzbadnixlWpdYD9Hnh2dDPKqIlDFS6StQLAuwxe48es0Vf51WFYFdzgTBbcW7m6BgpNrpQ2Cy6WFNDiopAHkFBeDlTZ0hlDqkrik6lnLDCojlRtnITG0Du7e8ghegR6Cu/KEN1ztwO0qhLsoVV5EqCFBgCR+wyktM+ADfPHhhsNoCr2eJMhW91z/qdV/OtcfNfpPF/r+obiTROjawYy0Q8lIhb4osjSYeqpfer3xJAE+p6PFyXZ/+ryASIOUkhhM22PS2v8WKYHeawc211uVwE7XILgCX0ejC9jmUuSQyRIo1dmOKiXFJeUjClfZb+ZNFbDUZgcF33lS4ddvD98f7u7jL5RWn56Sui48JsJLMu74fHHzOmJjNmHv2Hv2gX1kn9jnfRcHr50G3rYJviRJ20yhtyODSdRkKke5zh2Drcxc3todWoBHTXN1vG299OaWwXxldAmDwQCcjs5ckcbo5I12nUPOOtGqLorHtrSsyX/JN/hYcWnCIYNR1DKyWKBwYPQWhlSZecKGyWhBJVJX6tbv989Ld6DuqUTTS27jP3K7JDKKfpXRVTK5JEe0iBQbJZN/Jzb2xPY+s4RnkP6TApsOKOheoTq4nMyPtaM0U1vzpX5uz5LG7HppZ7JNrJe0abaN6KDVNeahBNfFtweEthMfC1y52783zb8TmtDvV0fGY7ae/qOXUzhXnPgBQVdyJQs0zSxZkABfUJBf0w2XwzntukkTCD9nHp2p1cYjFHJ5NNaV6+ZS0CciNzSFxsNgf+bnW+2q2v2QVi4LP5cIBvcLen8CkW8rEQ==).

I have no insights in the underlying architecture, but I already have **some comments on my PR** (and I am willing to work on these, if I get some feedback/hints):

* the **decompression happens synchronously** (`zlib.inflateSync`). This can be changed easily, once I understand how async operations are done for the compiler explorer software.
* the link is `/zclientstate/` (in addition to `/clientstate/`) - I am open for alternatives...
* I have **not added automated tests** (could be done, give me a hint/example)

The modified SW worked for me! I hope to get some feedback... **This would enable larger code-examples to be sent via clientstate as before with smaller URLs**.

mattgodbolt commented:
Hi! We've already supported compressed links for a while, though apparently not for clientstate on the server! On the client we use `static/url.ts`'s code and `lz-string` for lempel-ziv compression. It would be nice to use the same approach here: lz-string supports direct "decompressFromBase64".

That said: if that puts too high a burden on the creators, zlib is fine too!

As a general point:
- decompression happening synchronously: what other choice do we have? :)  There's no IO here to do asynchronously, and we don't have multiple threads
- I'd prefer auto-detecting the compressed stream rather than a new endpoint if that makes sense/is feasible. Failing that the endpoint `clientstatez` is preferable, thanks.
- Can you update the docs (docs/API) too with this new capability.
- Automated tests: if you split the state decompression into a separate part of the code instead of being inline in the route handler we can write a simple test without involving all the web framework stuff. happy to help with this.

goto40 commented:
I will update the PR. Thanks!

mattgodbolt commented:
CC @partouf too for their wisdom

goto40 commented:
@mattgodbolt I just tried to compress my link with `lzstring`

It seems that it is less performant than the zlib library (compressed in python, with `lzstring`):
* lzstring-compression: original length = 4012, compressed length= 2340
* zlib-compression (Level=9): original length = 4012, compressed length= 1652

side note: It was also much easier to integrate zlib, since it seems to be included as core module in node, as well as python - in contrast to lzstring.

**Current step: I would like to discuss if** 
* **solution 1**: we want to include lzstring as an option (in addition to zlib)
* solution 2: we want to just use zlib (an no lzstring) 
* solution 3: we want to just use lzstring (an no zlib) 

As far as I understood, solution 2 is no option, since it breaks with the current design. Solution 3 will be a bonus (smaller links), but will not solve my problem (my links will be still too large). So probably it will be **solution 1**. **What do you think?**

**Next step:** I have not tested/checked the documentation about the automatic detection of the format (compressed/uncompressed). I will check this, once we decided about solution 1-3 (but @partouf, if you have experience in that or some advice, any hint is appreciated).

Happy to hear from you (thank you for you interest in supporting me)!

mattgodbolt commented:
Given your results I'm happy to go with solution 2: let's not use lz-string if it's not better than the much-more-standard zlib :)

mattgodbolt commented:
I don't think zlib will break anything -- we keep lzstring in the static/url.ts case (we have to keep it for bckwards compatiblity). The server-side code you're working with will just remain separate from that; and we'll use zlib for the clientstate endpoint.

goto40 commented:
Please have a look at the updated PR (to test interactively, now use this new [link](http://localhost:10240/clientstate/eNq1VduK2zAQ/ZXBhWITJc6lVye7sBT6uIWl9CUOiyJPYm1syUjyZsOSf+/Idm5tWuhD7YCs0cyZM5doXgOL1kqtbJDA/DWQGa0jBkHB1brma6RtIHq9gERW10Y0glTFMYiqMrhCg0pgArlzlU3iGNXg/GAgdBlvYxLFQivHpUITl5mtuErVG6lEUWcIM6mtM8jL23PhAdLw7WAtXV4va4vGw6ByDfBGbzbadnixlWpdYD9Hnh2dDPKqIlDFS6StQLAuwxe48es0Vf51WFYFdzgTBbcW7m6BgpNrpQ2Cy6WFNDiopAHkFBeDlTZ0hlDqkrik6lnLDCojlRtnITG0Du7e8ghegR6Cu/KEN1ztwO0qhLsoVV5EqCFBgCR+wyktM+ADfPHhhsNoCr2eJMhW91z/qdV/OtcfNfpPF/r+obiTROjawYy0Q8lIhb4osjSYeqpfer3xJAE+p6PFyXZ/+ryASIOUkhhM22PS2v8WKYHeawc211uVwE7XILgCX0ejC9jmUuSQyRIo1dmOKiXFJeUjClfZb+ZNFbDUZgcF33lS4ddvD98f7u7jL5RWn56Sui48JsJLMu74fHHzOmJjNmHv2Hv2gX1kn9jnfRcHr50G3rYJviRJ20yhtyODSdRkKke5zh2Drcxc3todWoBHTXN1vG299OaWwXxldAmDwQCcjs5ckcbo5I12nUPOOtGqLorHtrSsyX/JN/hYcWnCIYNR1DKyWKBwYPQWhlSZecKGyWhBJVJX6tbv989Ld6DuqUTTS27jP3K7JDKKfpXRVTK5JEe0iBQbJZN/Jzb2xPY+s4RnkP6TApsOKOheoTq4nMyPtaM0U1vzpX5uz5LG7HppZ7JNrJe0abaN6KDVNeahBNfFtweEthMfC1y52783zb8TmtDvV0fGY7ae/qOXUzhXnPgBQVdyJQs0zSxZkABfUJBf0w2XwzntukkTCD9nHp2p1cYjFHJ5NNaV6+ZS0CciNzSFxsNgf+bnW+2q2v2QVi4LP5cIBvcLen8CkW8rEQ==).

I was a little surprised about the statement in the docs "To avoid problems in reading base64 by the API, some characters must be kept in unicode." (in docs/API.md). Do we have a problem with the compressed binary blog containing literally any possible byte value? I did not understand what exactly the problem is/was...

Once everything is OK so far, we could continue to add some tests.

mattgodbolt commented:
Oh you're not wrong, that's surprising to me too. @partouf do you recall what issue the docs are referring to? Seems to have come in #4114

mattgodbolt reviewed (commented):
I approve of this approach; thank you! Some tiny tests would be much appreciated; if you extract that whole anonymouse function to a free `export`ed function, then you could add some tests pretty easily: make a `test/handlers/route-api-test.ts` and follow along the approach of say something like `cfg-tests.ts`. You can abstract your anon function in terms of a `Buffer` so the tests just call with a Buffer

partouf commented:
> Oh you're not wrong, that's surprising to me too. @partouf do you recall what issue the docs are referring to? Seems to have come in #4114

It's some kind of artifact in json encoding, I don't really know why, but the entire internet knows about this and has workarounds for it

Example here:
https://github.com/compiler-explorer/compiler-explorer/blob/49aa8600974586ee1655c881c479e089bbfb7430/static/panes/editor.ts#L886

Which uses this:
https://github.com/compiler-explorer/compiler-explorer/blob/49aa8600974586ee1655c881c479e089bbfb7430/static/panes/editor.ts#L729

Fun fact, it's even here on a random dropbox api blurb:
https://www.dropbox.com/developers/reference/json-encoding

partouf commented:
Reading https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#well-formed_json.stringify I wonder if it might have been solved by the new "well-formed" json.stringify, but I remember testing this 3 years ago and I had the problems still then

goto40 commented:
@mattgodbolt I extracted a free function and added three test cases, one for each possible branch in the function:

* no compression
* no compression with first char indicating compression
* with compression

mattgodbolt commented on test/handlers/route-api-test.ts:
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2024, Pierre Bayerl
```
You added yourself to CONTRIBUTORS; do you really want to keep the copyright to yourself (which you can if you really want), or else put "Compiler Explorer Authors" like in the other files?

goto40 replied:
Thank you for this hint: I changed the text as you proposed (I originally just copied this copyright from some other file, where a name was provided).

goto40 replied:
I hope it was okay that I added myself to CONTRIBUTORS (I read this somewhere in the docs or while creating the PR). My main motivation is to get the compression feature, but of course I am also very proud to be able to contribute to this great project :-) - even if it is only a small contribution!

Furthermore, I would like to know - once we agreed on this PR - how the roll-out process looks like: when can I expect that this new feature is available online (to update my slides :-) )?

mattgodbolt replied:
Absolutely! This should be rolled out in a few hours (I'll do it at lunchtime!)

mattgodbolt reviewed (approved):
Thank you so much for this!

mattgodbolt commented:
This is rolling out to production and will be live in 30-40minutes. I tested it in our staging instance with:

https://godbolt.org/staging/clientstate/eNq1VduK2zAQ/ZXBhWITJc6lVye7sBT6uIWl9CUOiyJPYm1syUjyZsOSf+/Idm5tWuhD7YCs0cyZM5doXgOL1kqtbJDA/DWQGa0jBkHB1brma6RtIHq9gERW10Y0glTFMYiqMrhCg0pgArlzlU3iGNXg/GAgdBlvYxLFQivHpUITl5mtuErVG6lEUWcIM6mtM8jL23PhAdLw7WAtXV4va4vGw6ByDfBGbzbadnixlWpdYD9Hnh2dDPKqIlDFS6StQLAuwxe48es0Vf51WFYFdzgTBbcW7m6BgpNrpQ2Cy6WFNDiopAHkFBeDlTZ0hlDqkrik6lnLDCojlRtnITG0Du7e8ghegR6Cu/KEN1ztwO0qhLsoVV5EqCFBgCR+wyktM+ADfPHhhsNoCr2eJMhW91z/qdV/OtcfNfpPF/r+obiTROjawYy0Q8lIhb4osjSYeqpfer3xJAE+p6PFyXZ/+ryASIOUkhhM22PS2v8WKYHeawc211uVwE7XILgCX0ejC9jmUuSQyRIo1dmOKiXFJeUjClfZb+ZNFbDUZgcF33lS4ddvD98f7u7jL5RWn56Sui48JsJLMu74fHHzOmJjNmHv2Hv2gX1kn9jnfRcHr50G3rYJviRJ20yhtyODSdRkKke5zh2Drcxc3todWoBHTXN1vG299OaWwXxldAmDwQCcjs5ckcbo5I12nUPOOtGqLorHtrSsyX/JN/hYcWnCIYNR1DKyWKBwYPQWhlSZecKGyWhBJVJX6tbv989Ld6DuqUTTS27jP3K7JDKKfpXRVTK5JEe0iBQbJZN/Jzb2xPY+s4RnkP6TApsOKOheoTq4nMyPtaM0U1vzpX5uz5LG7HppZ7JNrJe0abaN6KDVNeahBNfFtweEthMfC1y52783zb8TmtDvV0fGY7ae/qOXUzhXnPgBQVdyJQs0zSxZkABfUJBf0w2XwzntukkTCD9nHp2p1cYjFHJ5NNaV6+ZS0CciNzSFxsNgf+bnW+2q2v2QVi4LP5cIBvcLen8CkW8rEQ==

goto40 commented:
Works :-)  All my links are <<2048 chars now :-)
